fathom/src/lang/core.rs

```diff
@@ -276,3 +280,145 @@ impl Default for Globals {
         Globals::new(entries)
     }
 }
+
+/// A [de Bruijn index][de-bruijn-index] in the local environment.
+///
+/// De bruijn indices describe an occurrence of a variable in terms of the
+/// number of binders between the occurrence and its associated binder.
+/// For example:
+///
+/// | Representation    | Example (S combinator)  |
+/// | ----------------- | ----------------------- |
+/// | Named             | `λx. λy. λz. x z (y z)` |
+/// | De Bruijn indices | `λ_. λ_. λ_. 2 0 (1 0)` |
+///
+/// This is a helpful representation because it allows us to easily compare
+/// terms for equivalence based on their binding structure without maintaining a
+/// list of name substitutions. For example we want `λx. x` to be the same as
+/// `λy. y`. With de Bruijn indices these would both be described as `λ 0`.
+///
+/// [de-bruijn-index]: https://en.wikipedia.org/wiki/De_Bruijn_index
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct LocalIndex(pub u32);
+
+impl LocalIndex {
+    /// Convert a local index to a local level in the current environment.
+    ///
+    /// `None` is returned if the local environment is not large enough to
+    /// contain the local variable.
+    pub fn to_level(self, size: LocalSize) -> Option<LocalLevel> {
+        Some(LocalLevel(u32::checked_sub(size.0, self.0 + 1)?))
+    }
+}
+
+/// A de Bruijn level in the local environment.
+///
+/// This describes an occurrence of a variable by counting the binders from
+/// the top of the term towards until the occurrence is reached.
+/// For example:
+///
+/// | Representation    | Example (S combinator)  |
+/// | ----------------- | ----------------------- |
+/// | Named             | `λx. λy. λz. x z (y z)` |
+/// | De Bruijn levels  | `λ_. λ_. λ_. 0 2 (1 2)` |
+///
+/// Levels are used in [values][semantics::Value] because they are not context-
```

Comment:
Interesting!  I was unaware of this technique :bulb:

Reply:
Yeah it's a super nifty way of doing dependent type checking

Reply:
> This describes an occurrence of a variable by counting the binders from the top of the term towards until the occurrence is reached.

Is there a word missing after towards?

Reply:
Good catch! Rewording this to:

> This describes an occurrence of a variable by counting the binders inwards from the top of the term until the occurrence is reached.